src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php

```diff
@@ -53,29 +54,67 @@ public function __construct(
      */
     public function handle(ListRootCategoriesWithCount $query): array
     {
-        $categorySelectedAsFilter = -1 !== $query->categoryIdSelectedAsFilter() ?
-            $this->categoryRepository->find($query->categoryIdSelectedAsFilter()) : null;
+        $defaultCategoryTreeId = $this->getDefaultCategoryTreeId($query);
+        if ($defaultCategoryTreeId === null) {
+            return [];
+        }
 
-        if (null === $categorySelectedAsFilter) {
-            $categorySelectedAsFilter = $this->userContext->getAccessibleUserTree();
-            if ($categorySelectedAsFilter === null) {
-                return [];
-            }
+        $selectedCategoryTree = $this->getCategoryTreeFromSelectedSubCategory(
+            $query->categoryIdSelectedAsFilter(),
+            $defaultCategoryTreeId
+        );
+
+        if ($selectedCategoryTree === null) {
+            return [];
         }
-        $rootCategoryIdToExpand = $categorySelectedAsFilter->getRoot();
 
-        $rootCategories = $query->countIncludingSubCategories() ?
+        return $query->countIncludingSubCategories() ?
             $this->listAndCountIncludingSubCategories->list(
                 $query->translationLocaleCode(),
                 $query->userId(),
-                $rootCategoryIdToExpand
+                $selectedCategoryTree->getId()
             ) :
             $this->listAndCountNotIncludingSubCategories->list(
                 $query->translationLocaleCode(),
                 $query->userId(),
-                $rootCategoryIdToExpand
+                $selectedCategoryTree->getId()
             );
+    }
+
+    /**
+     * Try to get the categoryTreeId from url parameters.
+     * If not set, try to get the root category id from selectedCategoryId in url parameters.
+     * If the root category does not exist, try to get the user accessible categoryTreeId.
+     * @param ListRootCategoriesWithCount $query
+     * @return int|null
+     */
+    private function getDefaultCategoryTreeId(ListRootCategoriesWithCount $query): ?int
+    {
+        $defaultCategoryTreeId = $this->userContext->getAccessibleUserTree()?->getId();
+        if (null !== $query->categoryTreeIdSelectedAsFilter()) {
+            $defaultCategoryTreeId = $query->categoryTreeIdSelectedAsFilter();
+        } else if (null !== $query->categoryIdSelectedAsFilter()) {
+            $selectedCategory = $this->categoryRepository->find($query->categoryIdSelectedAsFilter());
+            if (null !== $selectedCategory && null !== $selectedCategory->getRoot()) {
+                $defaultCategoryTreeId = $selectedCategory->getRoot();
+            }
+        }
+        return $defaultCategoryTreeId;
```

Comment:
Suggestion to use early return, easier to read & match the comment steps order.

```suggestion
    if (null !== $query->categoryTreeIdSelectedAsFilter()) {
        return $query->categoryTreeIdSelectedAsFilter();
    }

    if (null !== $query->categoryIdSelectedAsFilter()) {
        $selectedCategory = $this->categoryRepository->find($query->categoryIdSelectedAsFilter());
        if (null !== $selectedCategory && null !== $selectedCategory->getRoot()) {
            return $selectedCategory->getRoot();
        }
    }

    return $this->userContext->getAccessibleUserTree()?->getId();
```